lib/fluent/output.rb

```diff
@@ -291,7 +291,8 @@ def format_stream(tag, es)
     #end
 
     def enqueue_buffer(force = false)
-      @buffer.keys.each {|key|
+      buf_keys = @buffer.synchronize { @buffer.keys }
```

Comment:
Do `@buffer.keys.dup` in critical sections.
Your code iterates the object in `@buffer` itself, and doesn't protect anything in `synchronize`.